### PR TITLE
Graph Overview

### PR DESCRIPTION
**Pull request by Vane11ope**
gifrecord_2019-01-20_121930
gifrecord_2019-01-20_112737

### CODE REVIEW

**Comment by ITAYC0HEN** [comment]
Wow wow wow @Vane11ope, that looks so good! Very good job! 
Since it's a big tetaure, will take some time to dive into the code review - meantime, as we discussed, please add documentation "/*! /brief... /params... /returns...*/"

**Comment by XVilka** [comment]
Wow, this an eyecandy.

File: src/widgets/GraphWidget.cpp [hunk 1]
```diff
@@ -39,3 +42,74 @@ GraphWidget::GraphWidget(MainWindow *main, QAction *action) :
...
+}
+
+void GraphWidget::disableOverviewRect()
+{
+    overviewWidget->graphView->rangeRect = QRectF(0, 0, 0, 0);
+    overviewWidget->graphView->viewport()->update();
+}
+
+void GraphWidget::adjustOverview()
+{
```
Comment (xarkes) [inline src/widgets/GraphWidget.cpp]: You should check that overviewWidget isn't null here.

File: src/widgets/GraphWidget.cpp [hunk 2]
```diff
@@ -39,3 +42,74 @@ GraphWidget::GraphWidget(MainWindow *main, QAction *action) :
...
+    }
+    x *= curScale;
+    y *= curScale;
+    overviewWidget->graphView->rangeRect = QRectF(x + xoff, y + yoff, w, h);
+    overviewWidget->graphView->viewport()->update();
+}
+
+void GraphWidget::adjustGraph()
+{
+    int x1 = overviewWidget->graphView->horizontalScrollBar()->value();
```
Comment (xarkes) [inline src/widgets/GraphWidget.cpp]: Same here, check that overviewWidget is not null.

File: src/widgets/GraphWidget.cpp [hunk 3]
```diff
@@ -39,3 +42,74 @@ GraphWidget::GraphWidget(MainWindow *main, QAction *action) :
...
+        disconnect(graphView, SIGNAL(refreshBlock()), this, SLOT(adjustOverview()));
+        disconnect(graphView->horizontalScrollBar(), SIGNAL(valueChanged()), this, SLOT(adjustOverview()));
+        disconnect(graphView->verticalScrollBar(), SIGNAL(valueChanged()), this, SLOT(adjustOverview()));
+        disconnect(overviewWidget->graphView, SIGNAL(mouseMoved()), this, SLOT(adjustGraph()));
+        disableOverviewRect();
+    }
+}
+
+void GraphWidget::disableOverviewRect()
+{
```
Comment (xarkes) [inline src/widgets/GraphWidget.cpp]: `overviewWidget` might be null.

File: src/widgets/GraphWidget.cpp [hunk 4]
```diff
@@ -39,3 +42,74 @@ GraphWidget::GraphWidget(MainWindow *main, QAction *action) :
...
 
 GraphWidget::~GraphWidget() {}
+
+void GraphWidget::toggleOverview(bool visibility)
+{
+    if (visibility) {
+        connect(graphView, SIGNAL(refreshBlock()), this, SLOT(adjustOverview()));
+        connect(graphView->horizontalScrollBar(), SIGNAL(valueChanged()), this, SLOT(adjustOverview()));
+        connect(graphView->verticalScrollBar(), SIGNAL(valueChanged()), this, SLOT(adjustOverview()));
+        connect(overviewWidget->graphView, SIGNAL(mouseMoved()), this, SLOT(adjustGraph()));
```
Comment (xarkes) [inline src/widgets/GraphWidget.cpp]: `overviewWidget` might be null

File: src/widgets/OverviewView.h [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            for (const auto &line : lines) {
+                for (const auto &t : line) {
+                    result += t.text;
+                }
+            }
+            return result;
+        }
+    };
+
+    struct Instr {
```
Comment (xarkes) [inline src/widgets/OverviewView.h]: I think all the structs below are copy pasted from DisassemblyGraphWidget, right?
If so you should refactor it to have them in only one location.

File: src/widgets/OverviewWidget.cpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    this->graphView = new OverviewView(this);
+    this->setWidget(graphView);
+
+    QShortcut *toggle_shortcut = new QShortcut(widgetShortcuts["OverviewWidget"], main);
+    connect(toggle_shortcut, &QShortcut::activated, this, [ = ]() {
+            toggleDockWidget(true); 
+            main->updateDockActionChecked(action);
+    });
+
+    connect(this, &QDockWidget::visibilityChanged, this, [ = ](bool visibility) {
```
Comment (xarkes) [inline src/widgets/OverviewWidget.cpp]: Use RefreshDeferrer API.

File: src/widgets/OverviewView.cpp [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+    connect(shortcut_take_false, SIGNAL(activated()), this, SLOT(takeFalse()));
+
+    //Export Graph menu
+    mMenu->addSeparator();
+    actionExportGraph.setText(tr("Export Graph"));
+    mMenu->addAction(&actionExportGraph);
+    connect(&actionExportGraph, SIGNAL(triggered(bool)), this, SLOT(on_actionExportGraph_triggered()));
+
+    mMenu->addSeparator();
+    actionSyncOffset.setText(tr("Sync/unsync offset"));
```
Comment (xarkes) [inline src/widgets/OverviewView.cpp]: In my opinion, the overview must always be synchronized with one GraphView.

File: src/widgets/OverviewView.cpp [hunk 8]
```diff
@@ -0,0 +1,712 @@
...
+    current_scale = 1.0;
+    adjustSize(viewport()->size().width(), viewport()->size().height());
+    loadCurrentGraph();
+    viewport()->update();
+    adjustScale();
+}
+
+void OverviewView::loadCurrentGraph()
+{
+    TempConfig tempConfig;
```
Comment (xarkes) [inline src/widgets/OverviewView.cpp]: I think that all this logic as also copy pasted from DisassemblerGraphView, am I right?
We don't want to maintain code at two different location, so we should refactor this :)

File: src/widgets/OverviewView.cpp [hunk 9]
```diff
@@ -0,0 +1,712 @@
...
+        if (first_draw) {
+            showBlock(blocks[entry]);
+            first_draw = false;
+        }
+    }
+}
+
+void OverviewView::prepareGraphNode(GraphBlock &block)
+{
+    DisassemblyBlock &db = disassembly_blocks[block.entry];
```
Comment (xarkes) [inline src/widgets/OverviewView.cpp]: Probably same comment as above, is it copy pasted code?

File: src/widgets/OverviewView.cpp [hunk 10]
```diff
@@ -0,0 +1,712 @@
...
+    charHeight = metrics.height();
+    charOffset = 0;
+    if (mFontMetrics)
+        delete mFontMetrics;
+    mFontMetrics = new CachedFontMetrics(this, font());
+}
+
+void OverviewView::drawBlock(QPainter &p, GraphView::GraphBlock &block)
+{
+    p.setPen(Qt::black);
```
Comment (xarkes) [inline src/widgets/OverviewView.cpp]: Copy pasted?

File: src/widgets/OverviewView.cpp [hunk 11]
```diff
@@ -0,0 +1,712 @@
...
+        transition_dont_seek = false;
+        return;
+    }
+    seekLocal(to->entry);
+}
+
+
+void OverviewView::on_actionExportGraph_triggered()
+{
+    QStringList filters;
```
Comment (xarkes) [inline src/widgets/OverviewView.cpp]: Why is this here?
I don't think it is useful to have it in this widget.

**Comment by Vane11ope** [comment]
@xarkes yeah exactly. since i thought DisassemblerGraphView was too much for this, i chose not to inherit the class at first, but now i think otherwise. i will fix it, or more like clean it up. thanks :)

**Comment by xarkes** [comment]
@Vane11ope We discussed with @thestr4ng3r and probably the best would be that the overview pulls the `disassembly_block` map from the DisassmblerGraphView so everything is already computed.
I think all it has to do then is to add a scale factor.

**Comment by a1ext** [comment]
Very cool feature!, good job. 
One note: when I click on overwiew and keep holding left mouse button pressend and move mouse it doesn't follow the mouse. It would be very convenient if it will do so.

**Comment by Vane11ope** [comment]
done.

**Comment by Vane11ope** [comment]
Done with missing null checks.

**Comment by Vane11ope** [comment]
Null checks + RefreshDeferrer applied too
